### ISSUE TITLE
[🐛 Bug]: Some failed SauceLabs run result in Unnamed job descriptions

### ISSUE THREAD

**Issue by gabriel-kohen-by** [webdriverio/webdriverio#7609]
### WebdriverIO Version

7.7.4

### Node.js Version

14

### Mode

WDIO Testrunner

### Which capabilities are you using?

```typescript
import config from './wdio.shared.conf';

const isDebugMode = process.env.DEBUG;

config.user = process.env.SAUCE_USERNAME;
config.key = process.env.SAUCE_ACCESS_KEY;
config.bail = ("SAUCE_BAIL" in process.env)  ? process.env.SAUCE_BAIL : 1;
config.region = 'us';
const BUILD_PREFIX = process.env.BUILD_PREFIX ? `GitHub Actions-` : '';
const defaultBrowserSauceOptions = {
  build: `${BUILD_PREFIX}Luminate Portal Dashboard Mfe build-${new Date().toISOString()}`,
  screenResolution: '1600x1200',
  extendedDebugging: true,
  capturePerformance: true,
  timeZone: 'Chicago'
};
const defaultBrowserDesktopSauceOptions = {
  screenResolution: '1600x1200'
};
config.services?.push([
  'sauce',
  {
    sauceConnect: isDebugMode,
    setJobNameInBeforeSuite: true
  }
]);
config.mochaOpts= {...config.mochaOpts, bail: (config.bail === 1) }

config.waitforTimeout = 20000;

config.capabilities = [
  {
    browserName: 'chrome',
    browserVersion: 'latest',
    platformName: 'Windows 10',
    'sauce:options': {
      ...defaultBrowserSauceOptions,
      ...defaultBrowserDesktopSauceOptions
    }
  },
  {
    browserName: 'chrome',
    browserVersion: 'latest',
    platformName: 'macOS 10.15',
    'sauce:options': {
      ...defaultBrowserSauceOptions,
      ...defaultBrowserDesktopSauceOptions
    }
  },
  {
    browserName: 'MicrosoftEdge',
    browserVersion: 'latest',
    platformName: 'Windows 10',
    'sauce:options': {
      ...defaultBrowserSauceOptions,
      ...defaultBrowserDesktopSauceOptions
    }
  },
  {
    browserName: 'firefox',
    browserVersion: 'latest',
    platformName: 'macOS 10.15',
    'sauce:options': {
      ...defaultBrowserSauceOptions,
      ...defaultBrowserDesktopSauceOptions
    }
  }
];

export { config };
```

### What happened?

Some of our tests we are running fail in some of the capabilities. When these specific test fail on Chrome 95(or Windows 10 or Mac 10.15) or Firefox 93 on Mac 10.15 the description of the job is not set.
It's worth mentioning that in other tests that pass on the capabilities above and have no problems with the test description.
![image](https://user-images.githubusercontent.com/57095228/138904746-338cc7ca-5254-4638-b60d-7341c0183441.png)
Sometimes we will have successful test one Chrome 95/Windows 10 with an un-set job description:
![image](https://user-images.githubusercontent.com/57095228/138905095-c2eb67d5-cedc-4a8d-846a-df5debb1bcb6.png)

I've reached out to SauceLabs and was encouraged to reach out the his community. I'd love to contribute. Please let me know any starting points I can use to start diagnosing/fixing this. 

Thanks.

Shared config:
```
import getLogger from '@wdio/logger';
import 'expect-webdriverio';
import * as markdownReporter from 'wdio-markdown-reporter';
import { getReactTestIdComponent$, getReactTestIdComponent$$ } from '../commands/material-ui-commands';
import { getByTestId$, getByTestId$$ } from '../commands/test-id-commands';

const isDebugMode = process.env.DEBUG;

const log = getLogger('hooks');

const baseConfig: WebdriverIO.Config = {
  runner: 'local',
  specs: ['./e2e/specs/**/*.spec.ts'],
  exclude: [],
  maxInstances: isDebugMode ? 1 : 4,
  capabilities: [
    {
      browserName: 'chrome'
    }
  ],
  logLevel: isDebugMode ? 'debug' : 'info',
  outputDir: '.e2e/logs',
  bail: 0,
  baseUrl: process.env.LUMINATE_PORTAL_URL || 'http://localhost:7777/',
  waitforTimeout: 5000,
  waitforInterval: 500,
  connectionRetryTimeout: 30000,
  connectionRetryCount: 9,
  services: [],
  autoCompileOpts: {
    autoCompile: true,
    tsNodeOpts: {
      transpileOnly: true,
      project: './e2e/tsconfig.json'
    }
  },
  framework: 'mocha',
  reporters: [
    [
      markdownReporter,
      {
        outputDir: '.e2e/reports/markdown'
      }
    ],
    'spec'
  ],
  mochaOpts: {
    timeout: isDebugMode ? 60000 * 60 * 24 : 60000
  },

  before: function(capabilities, specs) {
    browser.addCommand('getReactTestIdComponent$', getReactTestIdComponent$);
    browser.addCommand('getReactTestIdComponent$$', getReactTestIdComponent$$);
    browser.addCommand('getByTestId$', getByTestId$);
    browser.addCommand('getByTestId$$', getByTestId$$);
    browser.addCommand('customConfig', function() {
      return {
        text: 'Hi Custom Command'
      };
    });
  },
  beforeSuite: function(suite) {
    log.debug(`Suite "${suite.fullTitle}" from file "${suite.file}" starts`);
  },
  beforeTest: function(test, context) {
    log.debug(`Test "${test.title}" starts`);
  },
  afterTest: function(test, context, { error, result, duration, passed, retries }) {
    log.debug(`Test "${test.title}" finished`);
  },
  afterSuite: function(suite) {
    log.debug(`Suite "${suite.fullTitle}" from file "${suite.file}" finished`);
  }
};

export default baseConfig;

```

### What is your expected behavior?

All test jobs should have their description populated according to the test runner framework(in out case it's Mocha).

### How to reproduce the bug.

Rerun the tests on Saucelabs.

### Relevant log output

```typescript
https://gist.github.com/gabriel-kohen-by/4edd984e0523ac5fabd0ca6e367754d4
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

**Comment by gabriel-kohen-by** [comment]
@christian-bromann , I've also noticed in many cases when running WDIO on Saucelabs, that specs say unnamed almost till the task is successful or failed(practically 70% of the run).

**Comment by christian-bromann** [comment]
The logs show that 4 jobs were updated:

```
2021-10-25T19:10:56.783Z INFO @wdio/sauce-service: Update job with sessionId 5287f207f12f47b1b0bf11965b0170d0, status: failing
2021-10-25T19:10:54.846Z INFO @wdio/sauce-service: Update job with sessionId 1f812d56699b4f1bb014e490764b2646, status: failing
2021-10-25T19:10:25.822Z INFO @wdio/sauce-service: Update job with sessionId df9d047a590540efba348173c74705cd, status: failing
2021-10-25T19:10:31.962Z INFO @wdio/sauce-service: Update job with sessionId 68fe6c1b6ad94135814638a43040e7ad, status: failing
```

There isn't much actionable for anyone here. Happy to add more logs to identify why it isn't updating the job properly.

**Comment by gabriel-kohen-by** [comment]
@christian-bromann , looked further and it seems like we're setting the job name according to the suite name on the beforeTest. Maybe we should set the job name on the beforeSuite since it will be the same suite name for every job on the suite.
Also worth noting that in runners such as Mocha, you'd have in the suite before step an action such as login. That will run before the current setting of the job name. That means that if it takes 20-30 seconds before you login to start your tests in the spec, you job name will show as unnamed job for quite a bit. Further, if that login step has failed, the job name will never be set. 
Should I create a PR to set the default job name to the suite on on beforeSuite?
In the time being I add my own:
```
  beforeSuite: function (suite) {
    browser.execute(`sauce:job-name=${suite.title}`);
  },
```

**Comment by christian-bromann** [comment]
> Further, if that login step has failed, the job name will never be set.

How so? The job update happens in the [`beforeTest`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-sauce-service/src/service.ts#L106) hook.

**Comment by gabriel-kohen-by** [comment]
Right, @christian-bromann. What happens if you have a Mocha `before` which runs at the beginning of each suite as a condition for any of the tests to run(and should not run again before every test). Login to a site is such an example.
That means that until that phase is done, the job (which includes the suite and all the tests) will not have it's name set.
Further, if there is an issue with login and it fails, the name will never be set for the job.

**Comment by christian-bromann** [comment]
Thanks for clarifying. I think your suggested solution seems most reasonable to me

> Should I create a PR to set the default job name to the suite on on beforeSuite?

**Comment by gabriel-kohen-by** [comment]
@christian-bromann thank you. Just to clarify, should I create a PR to set a default job name as a suite or should I keep using the suggested workaround?

**Comment by christian-bromann** [comment]
Ah right, a PR to set the name in `beforeSuite` seems fine to me

**Comment by gabriel-kohen-by** [comment]
@christian-bromann , thanks for the guidance. PR ready for review. Not sure if I should have added another test or what we have tests the criteria.

**Comment by christian-bromann** [comment]
fixed with #7672, thank you @gabriel-kohen-by for your contribution ❤️
